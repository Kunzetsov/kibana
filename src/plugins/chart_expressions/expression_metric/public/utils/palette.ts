/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { isColorDark } from '@elastic/eui';
import { Datatable } from 'src/plugins/expressions';

export type ColumnGroups<T = number[]> = Record<string, T>;

export interface MinMax {
  min: number;
  max: number;
}

export const getMinMax = (elements: number[]) => {
  if (!elements.length) {
    return { min: -Infinity, max: -Infinity };
  }

  const sortedArray = elements.sort((a, b) => a - b);
  return { min: sortedArray[0], max: sortedArray[sortedArray.length - 1] };
};

export const groupValuesByColumns = (datatable: Datatable) => {
  const numericColumns = datatable.columns
    .filter((column) => column.meta.type === 'number')
    .map((column) => column.id);

  return numericColumns.reduce<ColumnGroups>((groups, columnId) => {
    groups[columnId] = datatable.rows.map((row) => row[columnId]);
    return groups;
  }, {});
};

export const getMinMaxForColumns = (datatable: Datatable) => {
  const groups = groupValuesByColumns(datatable);
  return Object.keys(groups).reduce<ColumnGroups<MinMax>>((minMaxGroups, columnId) => {
    minMaxGroups[columnId] = getMinMax(groups[columnId]);
    return minMaxGroups;
  }, {});
};

export const parseRgbString = (rgb: string) => {
  const [group, r, g, b, opacityGroup, opacity] =
    rgb.match(/rgb\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*?(,\s*(\d+)\s*)?\)/) ?? [];

  return {
    red: parseFloat(r),
    green: parseFloat(g),
    blue: parseFloat(b),
    opacity: opacity ? parseFloat(opacity) : undefined,
  };
};

export const shouldApplyColor = (color: string) => {
  const { opacity } = parseRgbString(color);
  return opacity !== 0; // if opacity === 0 it means, there is no color to apply to the metric
};

export const needsLightText = (bgColor: string = '') => {
  const { red, green, blue, opacity } = parseRgbString(bgColor);
  return isColorDark(red, green, blue) && opacity !== 0;
};
