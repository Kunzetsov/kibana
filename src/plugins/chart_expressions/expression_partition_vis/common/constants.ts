/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export const PLUGIN_ID = 'expressionPartitionVis';
export const PLUGIN_NAME = 'expressionPartitionVis';

export const PIE_VIS_EXPRESSION_NAME = 'pie_vis';
export const PARTITION_VIS_RENDERER_NAME = 'partition_vis';
export const PARTITION_LABELS_VALUE = 'partition_labels_value';
export const PARTITION_LABELS_FUNCTION = 'partition_labels';

export const DEFAULT_PERCENT_DECIMALS = 2;
