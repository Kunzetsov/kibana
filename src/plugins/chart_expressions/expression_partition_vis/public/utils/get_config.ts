/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { PartitionConfig, PartitionLayout, RecursivePartial, Theme } from '@elastic/charts';
import { Datatable } from '../../../../../../src/plugins/expressions';
import {
  ChartTypes,
  LabelPositions,
  PartitionVisParams,
  PieContainerDimensions,
} from '../../common/types';

type Config = RecursivePartial<PartitionConfig>;

type GetConfigByTypeFn = (
  chartType: ChartTypes,
  visParams: PartitionVisParams,
  visData: Datatable,
  chartTheme: RecursivePartial<Theme>,
  dimensions?: PieContainerDimensions,
  rescaleFactor?: number
) => Config;

type GetConfigFn = (
  visParams: PartitionVisParams,
  visData: Datatable,
  chartTheme: RecursivePartial<Theme>,
  dimensions?: PieContainerDimensions,
  rescaleFactor?: number
) => Config;

const MAX_SIZE = 1000;

const getPieDonutWaffleCommonConfig: GetConfigFn = (
  visParams,
  visData,
  chartTheme,
  dimensions,
  rescaleFactor = 1
) => {
  const isSplitChart = Boolean(visParams.dimensions.splitColumn || visParams.dimensions.splitRow);
  const preventLinksFromShowing =
    (visParams.labels.position === LabelPositions.INSIDE || isSplitChart) && visParams.labels.show;
  const saveRescaleFactor = visParams.labels.show && !preventLinksFromShowing;

  const usingOuterSizeRatio =
    dimensions && !isSplitChart && !saveRescaleFactor
      ? {
          outerSizeRatio:
            // Cap the ratio to 1 and then rescale
            rescaleFactor * Math.min(MAX_SIZE / Math.min(dimensions?.width, dimensions?.height), 1),
        }
      : {};

  const config: Config = { ...usingOuterSizeRatio };

  if (!visParams.labels.show || visParams.labels.position === LabelPositions.HIDE) {
    // Force all labels to be linked, then prevent links from showing
    config.linkLabel = { maxCount: 0, maximumSection: Number.POSITIVE_INFINITY };
  }

  if (visParams.labels.last_level && visParams.labels.show) {
    config.linkLabel = {
      maxCount: Number.POSITIVE_INFINITY,
      maximumSection: Number.POSITIVE_INFINITY,
      maxTextLength: visParams.labels.truncate ?? undefined,
    };
  }

  if (preventLinksFromShowing) {
    config.linkLabel = { maxCount: 0 };
  }

  if (saveRescaleFactor) {
    // shrink up to 20% to give some room for the linked values
    config.outerSizeRatio = rescaleFactor;
  }

  return { ...config };
};

const getTreemapMosaicCommonConfig: GetConfigFn = (visParams) => {
  if (!visParams.labels.show || visParams.labels.position === LabelPositions.HIDE) {
    return {
      fillLabel: { textColor: 'rgba(0,0,0,0)' },
    };
  }
  return {};
};

const getPieSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.sunburst,
    ...getPieDonutWaffleCommonConfig(...args),
  };
};

const getDonutSpecificConfig: GetConfigFn = (visParams, ...args) => {
  return {
    partitionLayout: PartitionLayout.sunburst,
    emptySizeRatio: visParams.emptySizeRatio,
    ...getPieDonutWaffleCommonConfig(visParams, ...args),
  };
};

const getWaffleSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.waffle,
    ...getPieDonutWaffleCommonConfig(...args),
  };
};

const getTreemapSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.treemap,
    ...getTreemapMosaicCommonConfig(...args),
  };
};

const getMosaicSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.mosaic,
    ...getTreemapMosaicCommonConfig(...args),
  };
};

const getSpecificConfig: GetConfigByTypeFn = (chartType, ...args) =>
  ({
    [ChartTypes.PIE]: getPieSpecificConfig(...args),
    [ChartTypes.DONUT]: getDonutSpecificConfig(...args),
    [ChartTypes.TREEMAP]: getTreemapSpecificConfig(...args),
    [ChartTypes.MOSAIC]: getMosaicSpecificConfig(...args),
    [ChartTypes.WAFFLE]: getWaffleSpecificConfig(...args),
  }[chartType]);

export const getConfig: GetConfigByTypeFn = (
  chartType,
  visParams,
  visData,
  chartTheme,
  dimensions,
  rescaleFactor = 1
) => {
  // On small multiples we want the labels to only appear inside
  const isSplitChart = Boolean(visParams.dimensions.splitColumn || visParams.dimensions.splitRow);
  const usingMargin =
    dimensions && !isSplitChart
      ? {
          margin: {
            top: (1 - Math.min(1, MAX_SIZE / dimensions?.height)) / 2,
            bottom: (1 - Math.min(1, MAX_SIZE / dimensions?.height)) / 2,
            left: (1 - Math.min(1, MAX_SIZE / dimensions?.width)) / 2,
            right: (1 - Math.min(1, MAX_SIZE / dimensions?.width)) / 2,
          },
        }
      : null;
  return {
    fontFamily: chartTheme.barSeriesStyle?.displayValue?.fontFamily,
    outerSizeRatio: 1,
    specialFirstInnermostSector: false, // ?
    minFontSize: 10,
    maxFontSize: 16,
    emptySizeRatio: 0,
    sectorLineStroke: chartTheme.lineSeriesStyle?.point?.fill,
    sectorLineWidth: 1.5,
    circlePadding: 4,
    linkLabel: {
      maxCount: 5,
      fontSize: 11,
      textColor: chartTheme.axes?.axisTitle?.fill,
      maxTextLength: visParams.labels.truncate ?? undefined,
    },
    ...usingMargin,
    ...getSpecificConfig(chartType, visParams, visData, chartTheme, dimensions, rescaleFactor),
  };
};
