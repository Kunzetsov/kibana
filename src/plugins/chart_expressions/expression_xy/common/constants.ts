/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { i18n } from '@kbn/i18n';

export const XY_CHART = 'lens_xy_chart';
export const Y_CONFIG = 'lens_xy_yConfig';
export const MULTITABLE = 'lens_multitable';
export const DATA_LAYER = 'lens_xy_data_layer';
export const LEGEND_CONFIG = 'lens_xy_legendConfig';
export const XY_CHART_RENDERER = 'lens_xy_chart_renderer';
export const GRID_LINES_CONFIG = 'lens_xy_gridlinesConfig';
export const TICK_LABELS_CONFIG = 'lens_xy_tickLabelsConfig';
export const AXIS_EXTENT_CONFIG = 'lens_xy_axisExtentConfig';
export const REFERENCE_LINE_LAYER = 'lens_xy_referenceLine_layer';
export const LABELS_ORIENTATION_CONFIG = 'lens_xy_labelsOrientationConfig';
export const AXIS_TITLES_VISIBILITY_CONFIG = 'lens_xy_axisTitlesVisibilityConfig';

export const LayerTypes = {
  DATA: 'data',
  REFERENCELINE: 'referenceLine',
} as const;

export const FittingFunctions = {
  NONE: 'None',
  ZERO: 'Zero',
  LINEAR: 'Linear',
  CARRY: 'Carry',
  LOOKAHEAD: 'Lookahead',
} as const;

export const YAxisModes = {
  AUTO: 'auto',
  LEFT: 'left',
  RIGHT: 'right',
  BOTTOM: 'bottom',
} as const;

export const AxisExtentModes = {
  FULL: 'full',
  CUSTOM: 'custom',
  DATA_BOUNDS: 'dataBounds',
} as const;

export const LineStyles = {
  SOLID: 'solid',
  DASHED: 'dashed',
  DOTTED: 'dotted',
} as const;

export const FillStyles = {
  NONE: 'none',
  ABOVE: 'above',
  BELOW: 'below',
} as const;

export const IconPositions = {
  AUTO: 'auto',
  LEFT: 'left',
  RIGHT: 'right',
  ABOVE: 'above',
  BELOW: 'below',
} as const;

export const SeriesTypes = {
  BAR: 'bar',
  LINE: 'line',
  AREA: 'area',
  BAR_STACKED: 'bar_stacked',
  AREA_STACKED: 'area_stacked',
  BAR_HORIZONTAL: 'bar_horizontal',
  BAR_PERCENTAGE_STACKED: 'bar_percentage_stacked',
  BAR_HORIZONTAL_STACKED: 'bar_horizontal_stacked',
  AREA_PERCENTAGE_STACKED: 'area_percentage_stacked',
  BAR_HORIZONTAL_PERCENTAGE_STACKED: 'bar_horizontal_percentage_stacked',
} as const;

export const YScaleTypes = {
  TIME: 'time',
  LINEAR: 'linear',
  LOG: 'log',
  SQRT: 'sqrt',
} as const;

export const XScaleTypes = {
  TIME: 'time',
  LINEAR: 'linear',
  ORDINAL: 'ordinal',
} as const;

export const XYCurveTypes = {
  LINEAR: 'LINEAR',
  CURVE_MONOTONE_X: 'CURVE_MONOTONE_X',
} as const;

export const ValueLabelModes = {
  HIDE: 'hide',
  INSIDE: 'inside',
  OUTSIDE: 'outside',
} as const;

export const fittingFunctionDefinitions = [
  {
    id: FittingFunctions.NONE,
    title: i18n.translate('xpack.lens.fittingFunctionsTitle.none', {
      defaultMessage: 'Hide',
    }),
    description: i18n.translate('xpack.lens.fittingFunctionsDescription.none', {
      defaultMessage: 'Do not fill gaps',
    }),
  },
  {
    id: FittingFunctions.ZERO,
    title: i18n.translate('xpack.lens.fittingFunctionsTitle.zero', {
      defaultMessage: 'Zero',
    }),
    description: i18n.translate('xpack.lens.fittingFunctionsDescription.zero', {
      defaultMessage: 'Fill gaps with zeros',
    }),
  },
  {
    id: FittingFunctions.LINEAR,
    title: i18n.translate('xpack.lens.fittingFunctionsTitle.linear', {
      defaultMessage: 'Linear',
    }),
    description: i18n.translate('xpack.lens.fittingFunctionsDescription.linear', {
      defaultMessage: 'Fill gaps with a line',
    }),
  },
  {
    id: FittingFunctions.CARRY,
    title: i18n.translate('xpack.lens.fittingFunctionsTitle.carry', {
      defaultMessage: 'Last',
    }),
    description: i18n.translate('xpack.lens.fittingFunctionsDescription.carry', {
      defaultMessage: 'Fill gaps with the last value',
    }),
  },
  {
    id: FittingFunctions.LOOKAHEAD,
    title: i18n.translate('xpack.lens.fittingFunctionsTitle.lookahead', {
      defaultMessage: 'Next',
    }),
    description: i18n.translate('xpack.lens.fittingFunctionsDescription.lookahead', {
      defaultMessage: 'Fill gaps with the next value',
    }),
  },
];
