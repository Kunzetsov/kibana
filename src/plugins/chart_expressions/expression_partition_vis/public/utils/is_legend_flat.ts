/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { DatatableColumn } from '../../../../expressions';
import { ChartTypes } from '../../common/types';

type GetLegendIsFlatFn = (splitChartDimension: DatatableColumn | undefined) => boolean;

const isLegendFlatCommon: GetLegendIsFlatFn = (splitChartDimension) => Boolean(splitChartDimension);

const isLegendFlatWaffle: GetLegendIsFlatFn = (splitChartDimension) => true;

export const isLegendFlat = (
  visType: ChartTypes,
  splitChartDimension: DatatableColumn | undefined
) =>
  ({
    [ChartTypes.PIE]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.DONUT]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.TREEMAP]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.MOSAIC]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.WAFFLE]: isLegendFlatWaffle(splitChartDimension),
  }[visType]);
