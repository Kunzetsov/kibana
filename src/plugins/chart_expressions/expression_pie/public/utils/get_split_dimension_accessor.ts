/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import { AccessorFn } from '@elastic/charts';
import { getColumnByAccessor } from './accessor';
import { DatatableColumn } from '../../../../expressions/public';
import type { FieldFormatsStart } from '../../../../field_formats/public';
import { Dimension } from '../../common/types';

export const getSplitDimensionAccessor =
  (fieldFormats: FieldFormatsStart, columns: DatatableColumn[]) =>
  (splitDimension: Dimension): AccessorFn => {
    const formatter = fieldFormats.deserialize(splitDimension.format);
    const splitChartColumn = getColumnByAccessor(splitDimension.accessor, columns);
    const { id } = splitChartColumn ?? {};

    const fn: AccessorFn = (d) => {
      const v = d[accessor];
      if (v === undefined) {
        return;
      }
      const f = formatter.convert(v);
      return f;
    };

    return fn;
  };
