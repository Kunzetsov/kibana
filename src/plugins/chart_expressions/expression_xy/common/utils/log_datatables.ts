/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { ExecutionContext } from '../../../../expressions';
import { prepareLogTable } from '../../../../visualizations/common/utils';
import { LayerTypes } from '../constants';
import { strings } from '../i18n';
import { CommonXYLayerConfigResult } from '../types';

export const logDatatables = (layers: CommonXYLayerConfigResult[], handlers: ExecutionContext) => {
  if (!handlers?.inspectorAdapters?.tables) {
    return;
  }

  layers.forEach((layer, index) => {
    if (layer.layerType === LayerTypes.ANNOTATIONS) {
      return;
    }

    let xAccessor;
    let splitAccessor;
    if (layer.layerType === LayerTypes.DATA) {
      xAccessor = layer.xAccessor;
      splitAccessor = layer.splitAccessor;
    }

    const { accessors, layerType } = layer;
    const logTable = prepareLogTable(
      layer.table,
      [
        [
          accessors ? accessors : undefined,
          layerType === LayerTypes.DATA ? strings.getMetricHelp() : strings.getReferenceLineHelp(),
        ],
        [xAccessor ? [xAccessor] : undefined, strings.getXAxisHelp()],
        [splitAccessor ? [splitAccessor] : undefined, strings.getBreakdownHelp()],
      ],
      true
    );

    handlers.inspectorAdapters.tables.logDatatable(index, logTable);
  });
};
