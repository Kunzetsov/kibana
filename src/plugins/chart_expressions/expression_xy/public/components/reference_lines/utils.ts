/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Position } from '@elastic/charts';
import { IconPosition, YAxisMode } from '../../../common/types';

// if there's just one axis, put it on the other one
// otherwise use the same axis
// this function assume the chart is vertical
export function getBaseIconPlacement(
  iconPosition: IconPosition | undefined,
  axesMap?: Record<string, unknown>,
  axisMode?: YAxisMode
) {
  if (iconPosition === 'auto') {
    if (axisMode === 'bottom') {
      return Position.Top;
    }
    if (axesMap) {
      if (axisMode === 'left') {
        return axesMap.right ? Position.Left : Position.Right;
      }
      return axesMap.left ? Position.Right : Position.Left;
    }
  }

  if (iconPosition === 'left') {
    return Position.Left;
  }
  if (iconPosition === 'right') {
    return Position.Right;
  }
  if (iconPosition === 'below') {
    return Position.Bottom;
  }
  return Position.Top;
}
