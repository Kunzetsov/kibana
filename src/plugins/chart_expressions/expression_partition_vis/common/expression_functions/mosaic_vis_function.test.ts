/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { functionWrapper } from '../../../../expressions/common/expression_functions/specs/tests/utils';
import {
  MosaicVisConfig,
  LabelPositions,
  ValueFormats,
  LegendDisplay,
} from '../types/expression_renderers';
import { mosaicVisFunction } from './mosaic_vis_function';
import { Datatable } from '../../../../expressions/common/expression_types/specs';
import { PARTITION_LABELS_VALUE } from '../constants';

describe('interpreter/functions#mosaicVis', () => {
  const fn = functionWrapper(mosaicVisFunction());
  const context: Datatable = {
    type: 'datatable',
    rows: [{ 'col-0-1': 0 }],
    columns: [{ id: 'col-0-1', name: 'Count', meta: { type: 'number' } }],
  };

  const visConfig: MosaicVisConfig = {
    addTooltip: true,
    legendDisplay: LegendDisplay.SHOW,
    legendPosition: 'right',
    nestedLegend: true,
    truncateLegend: true,
    maxLegendLines: 2,
    palette: {
      type: 'system_palette',
      name: 'kibana_palette',
    },
    labels: {
      type: PARTITION_LABELS_VALUE,
      show: false,
      values: true,
      position: LabelPositions.DEFAULT,
      valuesFormat: ValueFormats.PERCENT,
      percentDecimals: 2,
      truncate: 100,
      lastLevel: false,
    },
    metric: {
      type: 'vis_dimension',
      accessor: 0,
      format: {
        id: 'number',
        params: {},
      },
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('returns an object with the correct structure', async () => {
    const actual = await fn(context, visConfig);
    expect(actual).toMatchSnapshot();
  });

  it('logs correct datatable to inspector', async () => {
    let loggedTable: Datatable;
    const handlers = {
      inspectorAdapters: {
        tables: {
          logDatatable: (name: string, datatable: Datatable) => {
            loggedTable = datatable;
          },
        },
      },
    };
    await fn(context, visConfig, handlers as any);

    expect(loggedTable!).toMatchSnapshot();
  });
});
