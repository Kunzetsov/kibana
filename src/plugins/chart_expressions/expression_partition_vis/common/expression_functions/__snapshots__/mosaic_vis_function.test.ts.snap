// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpreter/functions#mosaicVis logs correct datatable to inspector 1`] = `
Object {
  "columns": Array [
    Object {
      "id": "col-0-1",
      "meta": Object {
        "dimensionName": "Slice size",
        "type": "number",
      },
      "name": "Count",
    },
  ],
  "rows": Array [
    Object {
      "col-0-1": 0,
    },
  ],
  "type": "datatable",
}
`;

exports[`interpreter/functions#mosaicVis returns an object with the correct structure 1`] = `
Object {
  "as": "partitionVis",
  "type": "render",
  "value": Object {
    "params": Object {
      "listenOnChange": true,
    },
    "syncColors": false,
    "visConfig": Object {
      "addTooltip": true,
      "buckets": undefined,
      "dimensions": Object {
        "buckets": undefined,
        "metric": Object {
          "accessor": 0,
          "format": Object {
            "id": "number",
            "params": Object {},
          },
          "type": "vis_dimension",
        },
        "splitColumn": undefined,
        "splitRow": undefined,
      },
      "labels": Object {
        "lastLevel": false,
        "percentDecimals": 2,
        "position": "default",
        "show": false,
        "truncate": 100,
        "type": "partitionLabelsValue",
        "values": true,
        "valuesFormat": "percent",
      },
      "legendDisplay": "show",
      "legendPosition": "right",
      "maxLegendLines": 2,
      "metric": Object {
        "accessor": 0,
        "format": Object {
          "id": "number",
          "params": Object {},
        },
        "type": "vis_dimension",
      },
      "nestedLegend": true,
      "palette": Object {
        "name": "kibana_palette",
        "type": "system_palette",
      },
      "splitColumn": undefined,
      "splitRow": undefined,
      "truncateLegend": true,
    },
    "visData": Object {
      "columns": Array [
        Object {
          "id": "col-0-1",
          "meta": Object {
            "type": "number",
          },
          "name": "Count",
        },
      ],
      "rows": Array [
        Object {
          "col-0-1": 0,
        },
      ],
      "type": "datatable",
    },
    "visType": "mosaic",
  },
}
`;
